### PR TITLE
[Uid] Add fromBase58(), fromBase32(), fromRfc4122() and fromBinary() methods

### PR DESCRIPTION
**Pull request by fancyweb**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | https://github.com/symfony/symfony/issues/39801
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Component/Uid/AbstractUid.php [hunk 1]
```diff
@@ -37,6 +37,34 @@ abstract public static function isValid(string $uid): bool;
      */
     abstract public static function fromString(string $uid): self;
 
+    /**
+     * @return static
+     *
+     * @throws \InvalidArgumentException When the passed value is not valid
+     */
+    abstract public static function fromBinary(string $uid): self;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Uid/AbstractUid.php]: wouldn't it be possible to provide generic implementations?

File: src/Symfony/Component/Uid/Ulid.php [hunk 2]
```diff
@@ -58,14 +58,31 @@ public static function isValid(string $ulid): bool
      */
     public static function fromString(string $ulid): parent
     {
-        if (36 === \strlen($ulid) && Uuid::isValid($ulid)) {
-            $ulid = (new Uuid($ulid))->toBinary();
-        } elseif (22 === \strlen($ulid) && 22 === strspn($ulid, BinaryUtil::BASE58[''])) {
-            $ulid = BinaryUtil::fromBase($ulid, BinaryUtil::BASE58);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Uid/Ulid.php]: The implementation was correct and more efficient, I'd prefer to keep it if possible.

**Review by nicolas-grekas — COMMENTED** [review]
Thanks for starting this.

File: src/Symfony/Component/Uid/AbstractUid.php (hunk 1 above)
Comment (fancyweb) [reply src/Symfony/Component/Uid/AbstractUid.php]: I will have a look.

File: src/Symfony/Component/Uid/Ulid.php (hunk 2 above)
Comment (fancyweb) [reply src/Symfony/Component/Uid/Ulid.php]: Do you mean you want to keep all the current code in fromString() and duplicate the logic in the new methods?

Comment (nicolas-grekas) [reply src/Symfony/Component/Uid/Ulid.php]: I've been driven by the implementation here: implementing a selector-logic based on caught exceptions is extremely inefficient and not great from a design pov... :)

**Review by nicolas-grekas — APPROVED** [review]
(I updated the implementation to rely on a generic logic)

**Review by jderusse — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @fancyweb.

**Pull request merged**